tests/mainnet.rs

```diff
@@ -21,10 +21,10 @@ pub mod common;
 // TODO: [tokio::test]?
 #[test]
 fn mainnet_blocks_from_rpc() {
-    let rpc_url = match std::env::var("RPC_URL") {
+    let rpc_url = match std::env::var("ETHEREUM_RPC_URL") {
```

Comment:
We should also change `ci.yml`.

Reply:
Thanks. I have fixed it.

The CI here also fails with error 429. It means that `https://eth.llamarpc.com` is not the issue.

Let's merge this first: https://github.com/risechain/pevm/pull/333

Reply:
The CI in #333 also fails. It means that both #332 and #333 must be used to resolve the 429 errors.

I am merging the two PRs.

Reply:
This is great! Can you merge `self.runtime.block_on(with_retires(...))` into `RpcStorage::fetch` right below `new`? Then apply it to `block_hash` as well.

Reply:
> This is great! Can you merge `self.runtime.block_on(with_retires(...))` into `RpcStorage::fetch` right below `new`? Then apply it to `block_hash` as well.

You can have a look at the latest commit.

Personally, I recommend reverting this change because the code looks uglier. Note that the output of `self.provider.get_block_by_number` is not `Clone`.

Reply:
Based on my observations, the frequency of calls to `block_hash` is very rare. Therefore, it unlikely causes error 429.

Reply:
No, consistency is important. `block_hash` is an unlimited op-code and is frequently used for on-chain randomness. There's no reason to arbitrarily treat it as second-class.

Ugliness would only be a problem if we could write a function without any type parameter. Going from 2 to 4 while eliminating the `Clone` constraint with virtually no extra code lines is fine. Having random inconsistent logic is not.

Reply:
ok, you can merge this PR if everything looks good

Reply:
Ok. If readability is important let me add the missing `Clone` upstream. Personally, I find this pretty normal in Rust, and nowhere near Haskell's Monad Transformers and friends.